File: 2023/day-2/day_2.py
configuration_1 = {
        'red': 12,
        'green': 13,
        'blue': 14
}

def get_empty_totals():
      return { 'red': 0, 'green': 0, 'blue': 0 }

def get_color_and_quantity(selection):
      selection = selection.strip().split(' ')
      quantity = int(selection[0])
      color = selection[1]
      return color, quantity

def is_game_valid(game):
     maxes = configuration_1
     rounds = game.split(';')
     for round in rounds:
          totals = get_empty_totals()
          draws = round.split(',')
          for selection in draws:
               color, quantity = get_color_and_quantity(selection)
               totals[color] += quantity
               if totals[color] > maxes[color]:
                    return False
     return True

def get_game_power(game):
    total = 1
    maxes = get_empty_totals()
    rounds = game.split(';')
    for round in rounds:
         totals = get_empty_totals()
         draws = round.split(',')
         for selection in draws:
              color, quantity = get_color_and_quantity(selection)
              totals[color] += quantity
              maxes[color] = max(maxes[color], totals[color])
    
    for color in maxes.keys():
        total *= maxes[color]
    return maxes, total

def get_game_and_number(game):
    split_1 = game.split(':')
    game_num = int(split_1[0].split(' ')[1])
    return split_1[1], game_num

def part_one(input):
  valid_game_ids = []
  for line in input:
     game, num = get_game_and_number(line)
     if is_game_valid(game):
        valid_game_ids.append(num)
  return sum(valid_game_ids)

def part_two(input):
  result = 0
  for line in input:
     game, _ = get_game_and_number(line)
     _, power= get_game_power(game)  
     result += power
  return result

if __name__ == '__main__':
  with open('input.txt', 'r') as f:
          lines = f.readlines()
          input = [line.strip() for line in lines]

  print('part 1', part_one(input))
  print('part 2:', part_two(input))

File: 2023/day-2/day_2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var configuration1 = map[string]int{
	"red":   12,
	"green": 13,
	"blue":  14,
}

func getEmptyTotals() map[string]int {
	return map[string]int{"red": 0, "green": 0, "blue": 0}
}

func getColorAndQuantity(selection string) (string, int) {
	parts := strings.Fields(selection)
	quantity, _ := strconv.Atoi(parts[0])
	color := parts[1]
	return color, quantity
}

func isGameValid(game string) bool {
	maxes := configuration1
	rounds := strings.Split(game, ";")
	for _, round := range rounds {
		totals := getEmptyTotals()
		draws := strings.Split(round, ",")
		for _, selection := range draws {
			color, quantity := getColorAndQuantity(selection)
			totals[color] += quantity
			if totals[color] > maxes[color] {
				return false
			}
		}
	}
	return true
}

func getGamePower(game string) (map[string]int, int) {
	total := 1
	maxes := getEmptyTotals()
	rounds := strings.Split(game, ";")
	for _, round := range rounds {
		totals := getEmptyTotals()
		draws := strings.Split(round, ",")
		for _, selection := range draws {
			color, quantity := getColorAndQuantity(selection)
			totals[color] += quantity
			maxes[color] = max(maxes[color], totals[color])
		}
	}

	for _, value := range maxes {
		total *= value
	}
	return maxes, total
}

func getGameAndNumber(gameWithNum string) (string, int) {
	split1 := strings.Split(gameWithNum, ":")
	gameNum, _ := strconv.Atoi(strings.Fields(split1[0])[1])
	return split1[1], gameNum
}

func partOne(input []string) int {
	validGameIDs := []int{}
	for _, line := range input {
		game, num := getGameAndNumber(line)
		if isGameValid(game) {
			validGameIDs = append(validGameIDs, num)
		}
	}
	return sum(validGameIDs)
}

func partTwo(input []string) int {
	result := 0
	for _, line := range input {
		game, _ := getGameAndNumber(line)
		_, power := getGamePower(game)
		result += power
	}
	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func sum(nums []int) int {
	result := 0
	for _, num := range nums {
		result += num
	}
	return result
}

func readFile(filename string) []string {
	file, err := os.Open(filename)
	if err != nil {
		fmt.Println("Error opening file:", err)
		return []string{}
	}
	defer file.Close()

	var input []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		input = append(input, scanner.Text())
	}
	return input
}

func main() {
	input := readFile("input.txt")

	fmt.Println("part 1:", partOne(input))
	fmt.Println("part 2:", partTwo(input))
}
